### ISSUE TITLE
[Bug]: Person name + instance id is not a unique constraint

### ISSUE THREAD

**Issue by MrKaplan-lw** [LemmyNet/lemmy#5045]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Do you agree to follow the rules in our [Code of Conduct](https://join-lemmy.org/docs/code_of_conduct.html)?
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

Various places assume that the combination of `instance_id` (typically used as instance domain, but on DB level this would be the instance ID field) and username is a unique set, similar to the actor ID, but this is not actually enforced and can lead to duplicate `person`s in the DB.

### Steps to Reproduce

Unknown at this point, was likely due to incorrect configuration on the remote side generating HTTP URLs instead of HTTPS URLs for actor IDs temporarily.

### Technical Details

0.19.3 does not have a unique constraint for the combination of `instance_id` + `name` on the person table.
As far as I can tell this is the same in 0.19.5, as the list of indexes is the same for both versions:

```
Indexes:
    "person__pkey" PRIMARY KEY, btree (id)
    "idx_person_actor_id" UNIQUE CONSTRAINT, btree (actor_id)
    "idx_person_local_instance" btree (local DESC, instance_id)
    "idx_person_lower_actor_id" UNIQUE, btree (lower(actor_id::text))
    "idx_person_lower_name" btree (lower(name::text))
    "idx_person_published" btree (published DESC)
    "idx_person_trigram" gin (name gin_trgm_ops, display_name gin_trgm_ops)
```

```
lemmy=# select person.id, name, instance_id, actor_id from person join instance on instance.id = person.instance_id where name = 'Badabinski' and instance.domain = 'kbin.earth';
   id    |    name    | instance_id |            actor_id
---------+------------+-------------+---------------------------------
 9532845 | Badabinski |        6269 | http://kbin.earth/u/Badabinski
 8299094 | Badabinski |        6269 | https://kbin.earth/u/Badabinski
(2 rows)
```

Comments by the user are still showing up on lemmy.world, such as https://lemmy.world/comment/12501998 and also when looking at [the post](https://lemmy.world/post/20059287), but they are not associated with the `person` that is returned when looking at https://lemmy.world/u/Badabinski@kbin.earth.

See also https://lemmy.world/api/v3/user?person_id=8299094 and https://lemmy.world/api/v3/user?person_id=9532845.

### Version

0.19.3

### Lemmy Instance URL

lemmy.world

**Comment by Nutomic** [comment]
Most of these should be fixed by the url equality check I linked above. The remaining ones all seem to be caused by instances or users being deleted and then recreated. Though its also possible that a platform would allow multiple local users with the same name as long as they have different urls.

Can you point out where the assumption is made that (name, instance_id) is unique? That seems to be the real problem, but I cant think of any such case.

**Comment by Nutomic** [comment]
True, that is a simplification to improve usability. Lemmy itself has duplicate usernames when seen from other platforms, as you can have both actors /u/test and /c/test with the same name. [Activitypub](https://www.w3.org/TR/activitypub/) only specifies that the id url is unique (section 3.1), but the `preferredUsername` has no uniqueness guarantees (section 4.1).

The only solution I can think of would be to reference remote users by the actor_id, eg `/u/social.graves.cl/users/98vargwkp7` or using the db id like `/u/name-5124@social.graves.cl`. But both of those look ugly and will be confusing for new users.

**Comment by dessalines** [comment]
If this isn't possible in newer versions anyway, maybe instead of adding a constraint, we could do a migration that uses your query above to delete the duplicates (we've had to do migrations to do that before).

It could potentially be dangerous adding (person name + instance_id) uniqueness, since as @Nutomic said, person name doesn't have to be unique, only the actor_id.

**Comment by Nutomic** [comment]
A database constraint is not an ideal solution for federation because there can legitimately be multiple accounts with the same name and with their own post histories (eg instance deleted and recreated). It would then be impossible for these accounts to federate with Lemmy. However this is an uncommon case and we may decide not to support it. So then we would add the unique(name, instance_id) constraint and rename duplicate accounts.

This is essentially the same issue as https://github.com/LemmyNet/lemmy/issues/5052 which other Fediverse platforms experience with Lemmy, as Lemmy can have user and community with the same name. So to be consistent we should resolve that in the same way, by prohibiting creation of users and communities with the same name.
